File: cypha_som/gng_expert.py
"""Growing Neural Gas for auxiliary latent prototypes (Upgrade U1)."""

from __future__ import annotations

from typing import Dict, List, Optional, Tuple

import numpy as np


class GNGExpertManager:
    def __init__(
        self,
        d: int,
        eps_b: float = 0.05,
        eps_n: float = 0.006,
        lam: int = 100,
        age_max: int = 50,
        alpha_gng: float = 0.5,
        max_nodes: int = 256,
        rng: Optional[np.random.Generator] = None,
    ) -> None:
        self.d = int(d)
        self.eps_b = float(eps_b)
        self.eps_n = float(eps_n)
        self.lam = int(lam)
        self.age_max = int(age_max)
        self.alpha_gng = float(alpha_gng)
        self.max_nodes = int(max_nodes)
        self._rng = rng or np.random.default_rng(42)
        self._step_count = 0
        self._next_id = 0
        self.nodes: Dict[int, np.ndarray] = {}
        self.errors: Dict[int, float] = {}
        self.edges: Dict[Tuple[int, int], int] = {}
        self._init_two_nodes()

    def _init_two_nodes(self) -> None:
        for _ in range(2):
            nid = self._next_id
            self._next_id += 1
            self.nodes[nid] = self._rng.standard_normal(self.d) * 0.1
            self.errors[nid] = 0.0
        a, b = list(self.nodes.keys())[:2]
        self._add_edge(a, b)

    def _add_edge(self, i: int, j: int) -> None:
        if i == j:
            return
        key = (min(i, j), max(i, j))
        self.edges[key] = 0

    def _neighbors(self, i: int) -> List[int]:
        out: List[int] = []
        for (a, b), _ in self.edges.items():
            if a == i:
                out.append(b)
            elif b == i:
                out.append(a)
        return out

    def _two_closest(self, x: np.ndarray) -> Tuple[int, int]:
        x = np.asarray(x, dtype=np.float64).ravel()
        best_i, best_d = -1, float("inf")
        second_i, second_d = -1, float("inf")
        for nid, w in self.nodes.items():
            d = float(np.sum((x - w) ** 2))
            if d < best_d:
                second_i, second_d = best_i, best_d
                best_i, best_d = nid, d
            elif d < second_d:
                second_i, second_d = nid, d
        if second_i < 0:
            second_i = best_i
        return best_i, second_i

    def step(self, x: np.ndarray) -> int:
        x = np.asarray(x, dtype=np.float64).ravel()
        if x.size != self.d:
            raise ValueError(f"Expected dim {self.d}, got {x.size}")
        bmu, bmu2 = self._two_closest(x)
        self.nodes[bmu] += self.eps_b * (x - self.nodes[bmu])
        for nb in self._neighbors(bmu):
            self.nodes[nb] += self.eps_n * (x - self.nodes[nb])
        err = float(np.sum((x - self.nodes[bmu]) ** 2))
        self.errors[bmu] = self.errors.get(bmu, 0.0) + err
        self._add_edge(bmu, bmu2)
        key = (min(bmu, bmu2), max(bmu, bmu2))
        self.edges[key] = 0
        for ekey in list(self.edges.keys()):
            if bmu in ekey:
                self.edges[ekey] += 1
        self._prune_old_edges()
        self._step_count += 1
        if self._step_count % self.lam == 0 and len(self.nodes) < self.max_nodes:
            self._insert_node()
        self._decay_errors()
        return bmu

    def _prune_old_edges(self) -> None:
        dead = [k for k, age in self.edges.items() if age > self.age_max]
        for k in dead:
            del self.edges[k]
        isolated = [
            nid
            for nid in list(self.nodes.keys())
            if len(self._neighbors(nid)) == 0 and len(self.nodes) > 2
        ]
        for nid in isolated:
            self._remove_node(nid)

    def _remove_node(self, nid: int) -> None:
        self.nodes.pop(nid, None)
        self.errors.pop(nid, None)
        for k in [k for k in self.edges if nid in k]:
            del self.edges[k]

    def _insert_node(self) -> None:
        if not self.errors:
            return
        q = max(self.errors, key=lambda i: self.errors[i])
        nbrs = self._neighbors(q)
        if not nbrs:
            return
        f = max(nbrs, key=lambda j: self.errors.get(j, 0.0))
        r = self._next_id
        self._next_id += 1
        self.nodes[r] = 0.5 * (self.nodes[q] + self.nodes[f])
        self.errors[r] = self.errors[q]
        self.errors[q] *= self.alpha_gng
        self.errors[f] *= self.alpha_gng
        self._add_edge(q, r)
        self._add_edge(r, f)
        old = (min(q, f), max(q, f))
        if old in self.edges:
            del self.edges[old]

    def _decay_errors(self) -> None:
        for nid in self.errors:
            self.errors[nid] *= 0.995

    def force_insert(self, node_id: int) -> None:
        nbrs = self._neighbors(node_id)
        if not nbrs or node_id not in self.nodes:
            self._insert_node()
            return
        f = nbrs[0]
        r = self._next_id
        self._next_id += 1
        self.nodes[r] = 0.5 * (self.nodes[node_id] + self.nodes[f])
        self.errors[r] = self.errors.get(node_id, 1.0)
        self._add_edge(node_id, r)

    def merge_with_nearest(self, node_id: int) -> None:
        if node_id not in self.nodes:
            return
        nbrs = self._neighbors(node_id)
        if not nbrs:
            return
        other = nbrs[0]
        self.nodes[other] = 0.5 * (self.nodes[other] + self.nodes[node_id])
        self._remove_node(node_id)

    def node_count(self) -> int:
        return len(self.nodes)

    def get_prototypes(self) -> np.ndarray:
        if not self.nodes:
            return np.zeros((0, self.d), dtype=np.float64)
        return np.stack(list(self.nodes.values()), axis=0)

    def context_bias(self, h: np.ndarray, labels: List[str], strength: float = 0.15) -> Dict[str, float]:
        """Soft bias toward labels whose delta is nearest a GNG node (auxiliary routing)."""
        if not labels or not self.nodes:
            return {lbl: 0.0 for lbl in labels}
        h = np.asarray(h, dtype=np.float64).ravel()
        protos = self.get_prototypes()
        dists = np.linalg.norm(protos - h, axis=1)
        bmu = int(np.argmin(dists))
        # spread bias inversely with distance to all nodes
        weights = np.exp(-dists / (float(np.median(dists)) + 1e-9))
        weights /= weights.sum() + 1e-12
        out: Dict[str, float] = {}
        for i, lbl in enumerate(labels):
            idx = i % len(weights)
            out[lbl] = float(strength * weights[idx])
        return out

File: native/src/som/gng_expert.cpp
#include "cypha/som/gng_expert.hpp"

#include "cypha/numpy_default_rng.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace cypha::som {

namespace {

double sq_dist(const std::vector<double>& x, const std::vector<double>& w) {
  double s = 0.0;
  const int n = static_cast<int>(std::min(x.size(), w.size()));
  for (int i = 0; i < n; ++i) {
    const double d = x[static_cast<std::size_t>(i)] - w[static_cast<std::size_t>(i)];
    s += d * d;
  }
  for (std::size_t i = static_cast<std::size_t>(n); i < x.size(); ++i) {
    s += x[i] * x[i];
  }
  for (std::size_t i = static_cast<std::size_t>(n); i < w.size(); ++i) {
    s += w[i] * w[i];
  }
  return s;
}

}  // namespace

GNGExpertManager::GNGExpertManager(int d, GNGExpertConfig cfg)
    : d_(d),
      eps_b_(cfg.eps_b),
      eps_n_(cfg.eps_n),
      lam_(std::max(1, cfg.lam)),
      age_max_(cfg.age_max),
      alpha_gng_(cfg.alpha_gng),
      max_nodes_(std::max(2, cfg.max_nodes)),
      seed_(cfg.seed) {
  if (d_ <= 0) {
    throw std::invalid_argument("GNGExpertManager: d must be positive");
  }
  init_two_nodes();
}

void GNGExpertManager::init_two_nodes() {
  cypha::NumpyDefaultRng rng(static_cast<int>(seed_));
  nodes_.clear();
  errors_.clear();
  edges_.clear();
  next_id_ = 0;
  step_count_ = 0;

  for (int i = 0; i < 2; ++i) {
    const int nid = next_id_++;
    nodes_[nid] = std::vector<double>(static_cast<std::size_t>(d_), 0.0);
    for (int j = 0; j < d_; ++j) {
      nodes_[nid][static_cast<std::size_t>(j)] = rng.normal(0.0, 1.0) * 0.1;
    }
    errors_[nid] = 0.0;
  }
  auto it = nodes_.begin();
  const int a = it->first;
  ++it;
  const int b = it->first;
  add_edge(a, b);
}

void GNGExpertManager::add_edge(int i, int j) {
  if (i == j) {
    return;
  }
  const int lo = std::min(i, j);
  const int hi = std::max(i, j);
  edges_[{lo, hi}] = 0;
}

std::vector<int> GNGExpertManager::neighbors(int i) const {
  std::vector<int> out;
  for (const auto& kv : edges_) {
    if (kv.first.first == i) {
      out.push_back(kv.first.second);
    } else if (kv.first.second == i) {
      out.push_back(kv.first.first);
    }
  }
  return out;
}

std::pair<int, int> GNGExpertManager::two_closest(const std::vector<double>& x) const {
  int best_i = -1;
  double best_d = std::numeric_limits<double>::infinity();
  int second_i = -1;
  double second_d = std::numeric_limits<double>::infinity();
  for (const auto& kv : nodes_) {
    const double d = sq_dist(x, kv.second);
    if (d < best_d) {
      second_i = best_i;
      second_d = best_d;
      best_i = kv.first;
      best_d = d;
    } else if (d < second_d) {
      second_i = kv.first;
      second_d = d;
    }
  }
  if (second_i < 0) {
    second_i = best_i;
  }
  return {best_i, second_i};
}

void GNGExpertManager::prune_old_edges() {
  std::vector<std::pair<int, int>> dead;
  for (const auto& kv : edges_) {
    if (kv.second > age_max_) {
      dead.push_back(kv.first);
    }
  }
  for (const auto& k : dead) {
    edges_.erase(k);
  }

  std::vector<int> isolated;
  for (const auto& kv : nodes_) {
    if (neighbors(kv.first).empty() && static_cast<int>(nodes_.size()) > 2) {
      isolated.push_back(kv.first);
    }
  }
  for (int nid : isolated) {
    remove_node(nid);
  }
}

void GNGExpertManager::remove_node(int nid) {
  nodes_.erase(nid);
  errors_.erase(nid);
  std::vector<std::pair<int, int>> dead;
  for (const auto& kv : edges_) {
    if (kv.first.first == nid || kv.first.second == nid) {
      dead.push_back(kv.first);
    }
  }
  for (const auto& k : dead) {
    edges_.erase(k);
  }
}

void GNGExpertManager::insert_node() {
  if (errors_.empty()) {
    return;
  }
  int q = errors_.begin()->first;
  double q_err = errors_.begin()->second;
  for (const auto& kv : errors_) {
    if (kv.second > q_err) {
      q = kv.first;
      q_err = kv.second;
    }
  }
  const std::vector<int> nbrs = neighbors(q);
  if (nbrs.empty()) {
    return;
  }
  int f = nbrs[0];
  double f_err = errors_.count(f) ? errors_.at(f) : 0.0;
  for (int j : nbrs) {
    const double e = errors_.count(j) ? errors_.at(j) : 0.0;
    if (e > f_err) {
      f = j;
      f_err = e;
    }
  }

  const int r = next_id_++;
  nodes_[r] = std::vector<double>(static_cast<std::size_t>(d_), 0.0);
  for (int j = 0; j < d_; ++j) {
    nodes_[r][static_cast<std::size_t>(j)] =
        0.5 * (nodes_[q][static_cast<std::size_t>(j)] + nodes_[f][static_cast<std::size_t>(j)]);
  }
  errors_[r] = errors_[q];
  errors_[q] *= alpha_gng_;
  errors_[f] *= alpha_gng_;
  add_edge(q, r);
  add_edge(r, f);
  const int lo = std::min(q, f);
  const int hi = std::max(q, f);
  edges_.erase({lo, hi});
}

void GNGExpertManager::decay_errors() {
  for (auto& kv : errors_) {
    kv.second *= 0.995;
  }
}

int GNGExpertManager::step(const std::vector<double>& x) {
  if (static_cast<int>(x.size()) != d_) {
    throw std::invalid_argument("GNGExpertManager::step: dimension mismatch");
  }

  const auto [bmu, bmu2] = two_closest(x);

  for (int j = 0; j < d_; ++j) {
    nodes_[bmu][static_cast<std::size_t>(j)] +=
        eps_b_ * (x[static_cast<std::size_t>(j)] - nodes_[bmu][static_cast<std::size_t>(j)]);
  }
  for (int nb : neighbors(bmu)) {
    for (int j = 0; j < d_; ++j) {
      nodes_[nb][static_cast<std::size_t>(j)] +=
          eps_n_ * (x[static_cast<std::size_t>(j)] - nodes_[nb][static_cast<std::size_t>(j)]);
    }
  }

  const double err = sq_dist(x, nodes_[bmu]);
  errors_[bmu] = (errors_.count(bmu) ? errors_[bmu] : 0.0) + err;

  add_edge(bmu, bmu2);
  const int lo = std::min(bmu, bmu2);
  const int hi = std::max(bmu, bmu2);
  edges_[{lo, hi}] = 0;

  std::vector<std::pair<int, int>> touch;
  for (const auto& kv : edges_) {
    if (kv.first.first == bmu || kv.first.second == bmu) {
      touch.push_back(kv.first);
    }
  }
  for (const auto& k : touch) {
    edges_[k] += 1;
  }

  prune_old_edges();
  step_count_ += 1;
  if (step_count_ % lam_ == 0 && static_cast<int>(nodes_.size()) < max_nodes_) {
    insert_node();
  }
  decay_errors();
  return bmu;
}

std::vector<std::vector<double>> GNGExpertManager::get_prototypes() const {
  std::vector<std::vector<double>> out;
  out.reserve(nodes_.size());
  for (const auto& kv : nodes_) {
    out.push_back(kv.second);
  }
  return out;
}

void GNGExpertManager::force_insert(int node_id) {
  const std::vector<int> nbrs = neighbors(node_id);
  if (nbrs.empty() || nodes_.count(node_id) == 0) {
    insert_node();
    return;
  }
  const int f = nbrs[0];
  const int r = next_id_++;
  nodes_[r] = std::vector<double>(static_cast<std::size_t>(d_), 0.0);
  for (int j = 0; j < d_; ++j) {
    nodes_[r][static_cast<std::size_t>(j)] =
        0.5 * (nodes_[node_id][static_cast<std::size_t>(j)] + nodes_[f][static_cast<std::size_t>(j)]);
  }
  errors_[r] = errors_.count(node_id) ? errors_.at(node_id) : 1.0;
  add_edge(node_id, r);
}

void GNGExpertManager::merge_with_nearest(int node_id) {
  if (nodes_.count(node_id) == 0) {
    return;
  }
  const std::vector<int> nbrs = neighbors(node_id);
  if (nbrs.empty()) {
    return;
  }
  const int other = nbrs[0];
  for (int j = 0; j < d_; ++j) {
    nodes_[other][static_cast<std::size_t>(j)] =
        0.5 * (nodes_[other][static_cast<std::size_t>(j)] +
               nodes_[node_id][static_cast<std::size_t>(j)]);
  }
  remove_node(node_id);
}

}  // namespace cypha::som
